spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,308 @@
+        RichEqualTo(index, colDataType, v)
+      case LessThan(name, v) =>
+        val (index, colDataType) = indexAndType(name)
+        RichLessThan(index, colDataType, v)
+      case LessThanOrEqual(name, v) =>
+        val (index, colDataType) = indexAndType(name)
+        RichLessThanOrEqual(index, colDataType, v)
+      case GreaterThan(name, v) =>
+        val (index, colDataType) = indexAndType(name)
+        RichGreaterThan(index, colDataType, v)
+      case GreaterThanOrEqual(name, v) =>
+        val (index, colDataType) = indexAndType(name)
+        RichGreaterThanOrEqual(index, colDataType, v)
+      case IsNotNull(name) =>
+        val (index, colDataType) = indexAndType(name)
+        RichIsNotNull(index)
+      case IsNull(name) =>
+        val (index, colDataType) = indexAndType(name)
+        RichIsNull(index)
+      case In(name, v) =>
+        val (index, colDataType) = indexAndType(name)
+        RichIn(index, v)
+      case And(l, r) => And(richFilter(l), richFilter(r))
+      case Or(l, r) => Or(richFilter(l), richFilter(r))
+      case Not(f) => Not(richFilter(f))
+      case _ => throw new IllegalArgumentException(s"Unknown filter $filter")
+    }
+
+    /** Evaluate rich filter on a column value from a row */
+    def translateFilter(filter: Filter): Row => Boolean = filter match {
+      case RichEqualTo(colIndex, colDataType,  value) =>
+        (row: Row) => equalTo(row.get(colIndex), colDataType, value)
+      case RichLessThan(colIndex, colDataType, value) =>
+        (row: Row) => compare(row.get(colIndex), colDataType, value) < 0
+      case RichLessThanOrEqual(colIndex, colDataType, value) =>
+        (row: Row) => compare(row.get(colIndex), colDataType, value) <= 0
+      case RichGreaterThan(colIndex, colDataType, value) =>
+        (row: Row) => compare(row.get(colIndex), colDataType, value) > 0
+      case RichGreaterThanOrEqual(colIndex, colDataType, value) =>
+        (row: Row) => compare(row.get(colIndex), colDataType, value) >= 0
+      case RichIn(colIndex, values) => (row: Row) => values.toSet.contains(row.get(colIndex))
+      case RichIsNull(colIndex) => (row: Row) => row.get(colIndex) == null
+      case RichIsNotNull(colIndex) => (row: Row) => row.get(colIndex) != null
+      case Not(f) => (row: Row) => !translateFilter(f)(row)
+      case And(l, r) => (row: Row) => translateFilter(l)(row) && translateFilter(r)(row)
+      case Or(l, r) => (row: Row) => translateFilter(l)(row) || translateFilter(r)(row)
+      case _ => (row: Row) => throw new RuntimeException(s"Unknown RichFilter $filter")
+    }
+
+    // a filter combining all other filters
+    val richFilters = preservedFilters.map(richFilter)
+    val translatedFilters = richFilters.map(translateFilter)
+    def rowFilter(row: Row): Boolean = translatedFilters.forall(_(row))
+
+    val transformer = new RddTransformer(requiredColumns, pushdownFilters)
+    transformer.transform(baseRdd).filter(rowFilter)
```

Comment:
Calling `filter` on the result RDD is not needed here, because Spark docs says explicitly, Spark applies the filters by itself. There is no need to rewrite parts of Spark here. And once filter is not needed, 90% of the logic of this class can be removed - that is, all lines between L163 and L271. To double check, I removed these lines and it didn't cause any unit nor integration tests to fail.